### PR TITLE
implement ability to set properties in conn-string

### PR DESCRIPTION
**Pull request by mikethebeer**
JDBC client properties can be set via connection string
e.g. `crate://localhost:4300?property1=value1&...`

### CODE REVIEW

File: docs/index.txt [hunk 1]
```diff
@@ -134,6 +135,10 @@ Properties can be specified when connecting to Crate using the JDBC driver::
     properties.put(<key>, <value>);
     Connection conn = DriverManager.getConnection("crate://localhost:4300", properties);
 
+In addition client properties can be passed via the JDBC URL::
```
Comment (kovrus) [inline docs/index.txt]: i think it is called `connection properties or connection parameters`

File: src/main/java/io/crate/client/jdbc/CrateDriver.java [hunk 2]
```diff
@@ -66,23 +66,48 @@ public Connection connect(String url, Properties info) throws SQLException {
...
+                if (keyVal.length == 2) {
+                    properties.setProperty(keyVal[0], keyVal[1]);
+                } else {
+                    throw new InvalidPropertiesFormatException("Properties format is invalid. " +
+                            "Valid format is: property=value&property=value,...");
+                }
+            }
+            return properties;
+        } else {
+            return null;
```
Comment (kovrus) [inline src/main/java/io/crate/client/jdbc/CrateDriver.java]: do not return null, but instead return empty Properties obj.

File: src/main/java/io/crate/client/jdbc/CrateDriver.java [hunk 3]
```diff
@@ -66,23 +66,48 @@ public Connection connect(String url, Properties info) throws SQLException {
...
-                throw new SQLException("URL format is invalid. " +
-                        "Valid format is: [jdbc:]crate://[host1:port1][, host2:port2 ...]/[schema][?param=value]");
+                throw new SQLException(e.getMessage());
             }
         }
         return connection;
     }
 
-
+    private Properties parseProperties(String paramList) throws InvalidPropertiesFormatException {
```
Comment (kovrus) [inline src/main/java/io/crate/client/jdbc/CrateDriver.java]: maybe `paramList -> params`?

File: src/main/java/io/crate/client/jdbc/CrateConnection.java [hunk 4]
```diff
@@ -333,22 +335,28 @@ public boolean isValid(int timeout) throws SQLException {
...
+        if (value != null) {
+            this.properties.setProperty(name, value);
+        } else {
+            this.properties.remove(name);
+        }
     }
 
     @Override
     public void setClientInfo(Properties properties) throws SQLClientInfoException {
+        this.properties = properties;
```
Comment (mikethebeer) [inline src/main/java/io/crate/client/jdbc/CrateConnection.java]: clear all properties if empty list is provided

File: src/main/java/io/crate/client/jdbc/CrateDriver.java (hunk 3 above)
Comment (chaudum) [inline src/main/java/io/crate/client/jdbc/CrateDriver.java]: isn't there a utility function doing exactly this thing?

File: src/main/java/io/crate/client/jdbc/CrateDriver.java [hunk 5]
```diff
@@ -66,23 +66,48 @@ public Connection connect(String url, Properties info) throws SQLException {
...
-            String[] urlParts = url.split("/");
+            String[] pathAndParams = url.split("\\?");
             connection = new CrateConnection(handle, info);
             connection.connect();
 
-            if (urlParts.length == 2) {
-                String[] pathAndParams = urlParts[1].split("\\?");
-                connection.setSchema(pathAndParams[0]);
-            } else if (urlParts.length > 2) {
+            try {
```
Comment (chaudum) [inline src/main/java/io/crate/client/jdbc/CrateDriver.java]: only wrap necessary code into try catch

File: src/main/java/io/crate/client/jdbc/CrateConnection.java [hunk 6]
```diff
@@ -38,11 +38,13 @@
     private CrateDatabaseMetaData metaData;
     private String databaseVersion;
     private boolean strict;
+    private Properties properties;
 
     public CrateConnection(ClientHandleRegistry.ClientHandle handle, Properties properties) {
         this.clientHandle = handle;
         this.readOnly = false;
         this.strict = Boolean.valueOf(properties.getProperty("strict", "false"));
```
Comment (chaudum) [inline src/main/java/io/crate/client/jdbc/CrateConnection.java]: `strict` is also a property and should be handled the same way all others are handled

File: src/main/java/io/crate/client/jdbc/CrateDriver.java [hunk 7]
```diff
@@ -59,30 +64,84 @@ public Connection connect(String url, Properties info) throws SQLException {
...
+                    throw new InvalidPropertiesFormatException("Properties format is invalid. " +
+                            "Valid format is: property=value&property=value,...");
+                }
+            }
+        }
+
+        return url;
+    }
 
+    private void parseProperties(String paramList, Properties info) throws InvalidPropertiesFormatException {
```
Comment (mikethebeer) [inline src/main/java/io/crate/client/jdbc/CrateDriver.java]: not used anymore

File: src/main/java/io/crate/client/jdbc/CrateConnection.java [hunk 8]
```diff
@@ -333,22 +341,30 @@ public boolean isValid(int timeout) throws SQLException {
...
-        checkClosed();
-        return null;
+        return this.properties.getProperty(name);
     }
 
     @Override
     public Properties getClientInfo() throws SQLException {
-        checkClosed();
-        return null;
+        return this.properties;
```
Comment (chaudum) [inline src/main/java/io/crate/client/jdbc/CrateConnection.java]: `return properties`

File: src/main/java/io/crate/client/jdbc/CrateConnection.java [hunk 9]
```diff
@@ -333,22 +341,30 @@ public boolean isValid(int timeout) throws SQLException {
...
+        if (properties.isEmpty()) {
+            this.properties.clear();
+        } else {
+            this.properties = properties;
+        }
     }
 
     @Override
     public String getClientInfo(String name) throws SQLException {
-        checkClosed();
```
Comment (chaudum) [inline src/main/java/io/crate/client/jdbc/CrateConnection.java]: do not remove `checkClosed()`

File: src/main/java/io/crate/client/jdbc/CrateConnection.java [hunk 10]
```diff
@@ -333,22 +341,30 @@ public boolean isValid(int timeout) throws SQLException {
...
-        return null;
+        return this.properties.getProperty(name);
     }
 
     @Override
     public Properties getClientInfo() throws SQLException {
-        checkClosed();
```
Comment (chaudum) [inline src/main/java/io/crate/client/jdbc/CrateConnection.java]: do not remove `checkClosed()`

File: src/main/java/io/crate/client/jdbc/CrateConnection.java [hunk 11]
```diff
@@ -333,22 +341,30 @@ public boolean isValid(int timeout) throws SQLException {
...
+        } else {
+            this.properties = properties;
+        }
     }
 
     @Override
     public String getClientInfo(String name) throws SQLException {
-        checkClosed();
-        return null;
+        return this.properties.getProperty(name);
```
Comment (chaudum) [inline src/main/java/io/crate/client/jdbc/CrateConnection.java]: is it ok to return `null` if the property does not exists or should it throw an exception? pls check spec

File: src/test/java/io/crate/client/jdbc/CrateConnectionTest.java [hunk 12]
```diff
@@ -107,4 +108,52 @@ public void testClosedThreadpool() throws Exception {
...
+        conn.setClientInfo(new Properties());
+        assertThat(conn.getClientInfo().size(), is(0));
+
+        conn.setClientInfo("prop1", "value1");
+        conn.setClientInfo("prop2", "value2");
+        conn.setClientInfo("prop1", null);
+        assertThat(conn.getClientInfo().size(), is(1));
+
+        conn.close();
+        try {
```
Comment (chaudum) [inline src/test/java/io/crate/client/jdbc/CrateConnectionTest.java]: I would put that test case into a separate test method.

File: src/test/java/io/crate/client/jdbc/CrateConnectionTest.java [hunk 13]
```diff
@@ -107,4 +108,52 @@ public void testClosedThreadpool() throws Exception {
...
+
+        Properties properties = new Properties();
+        properties.setProperty("prop1", "value1");
+        properties.setProperty("prop2", "value2");
+        conn.setClientInfo(properties);
+        assertThat(conn.getClientInfo("prop2"), is("value2"));
+        assertThat(conn.getClientInfo(), is(properties));
+
+        conn.close();
+        try {
```
Comment (chaudum) [inline src/test/java/io/crate/client/jdbc/CrateConnectionTest.java]: same here

File: CHANGES.txt [hunk 14]
```diff
@@ -5,6 +5,8 @@ Changes for Crate Data JDBC Client
 Unreleased
 ==========
 
+ - Implemented ability to set JDBC client properties via the connection string.
```
Comment (chaudum) [inline CHANGES.txt]: suggestion: It is now possible to set connection properties via the connection string

**Comment by chaudum** [comment]
lgtm. @mfussenegger what do you think?

File: src/test/java/io/crate/client/jdbc/integrationtests/CrateJDBCDriverTest.java [hunk 15]
```diff
@@ -192,4 +194,35 @@ public void run() {
...
+        Connection conn = DriverManager.getConnection("crate://" + hostAndPort + "?prop1=value1&prop2=value2");
+        Properties properties = conn.getClientInfo();
+        assertThat(properties.size(), is(3));
+        assertThat(properties.getProperty("prop1"), is("value1"));
+        assertThat(properties.getProperty("prop2"), is("value2"));
+        conn.close();
+    }
+
+    @Test
+    public void testInvalidUrlAndPropertyFormat() throws Exception {
```
Comment (seut) [inline src/test/java/io/crate/client/jdbc/integrationtests/CrateJDBCDriverTest.java]: pls move to an unit test

**Pull request merged**